### DISCUSSION TITLE
Completion provider wired node hierarchy

### DISCUSSION THREAD

**Discussion by cdietrich** [eclipse-langium/langium#1279] [Q&A]
Hi, what is the background on this code

<img width="801" alt="Bildschirmfoto 2023-11-08 um 21 04 03" src="https://github.com/eclipse-langium/langium/assets/238736/3a7fec92-94c5-44fd-838e-353b2db6c275">

how can i detect this situation in the scope provider.
the type of the node and the parent node is the same with this in our grammar

(after langium 2.1.1. update)

grammar

```
grammar HelloWorld

entry Model:
    (content+=FileContent)*;

EnumLiteral:
    name=ID ','?;

EnumDeclaration:
    'enum' name=ID '{' (items+=EnumLiteral)*'}';

ValidationBlock: 'validations' '{' (content+=When)* '}';

When: 'when' condition=Expression '{' '}';

FileContent:
    ValidationBlock | EnumDeclaration;

Expression:
    Equality;

Equality infers Expression:
    DotExpression (({infer Equals.left = current} '==' | {infer NotEquals.left=current} '!=')  right=DotExpression)*;


DotExpression infers Expression:
    Primary ({infer DotExpression.base = current} '.' target=DotTarget)*;

Primary infers Expression:
    {infer Parenthesis} '(' expr=Expression ')' |
    RefExpression |
    NumberLiteral;

RefExpression:
    target=[NamedRefExpressionTarget:ID];

DotTarget:
   {infer DotTargetRef} ref=[NamedDotTargetReference:ID];

NamedRefExpressionTarget: EnumDeclaration;

NamedDotTargetReference: EnumLiteral;

NumberLiteral : value = NUMBER;

NUMBER returns number: NUMBER_LIT | 'NaN';
ID returns string: RawId | EscapedId;

hidden terminal WS: /\s+/;
terminal NUMBER_LIT returns number: /[0-9]+(\.[0-9]+)?/;
terminal RawId: /[_a-zA-Z][\w_-]*/;
terminal EscapedId: /`[^`]*`/;
terminal STRING: /"(\\.|[^"\\])*"|'(\\.|[^'\\])*'/;

hidden terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;
```

test model

```
validations {
        when(1 == M){  }
    }
enum ModelVisibility {
    public
    private
}

```

**Comment by msujew** [comment]
The idea behind creating a synthetic node at this point is that we might encounter a cross reference directly at the start of a parser rule. In this case, the AST node that is being picked doesn't match the node that we need to compute the scope for the cross reference. To do that, we need to construct a new object with the expected type.

Can you test whether this is also the case with `langium@2.1.2`? It released with https://github.com/eclipse-langium/langium/pull/1267 and should improve the accuracy which AST node is picked here.

Note that the AST picking logic might still be riddled with edge case issues, so a reproducible example would be great so we can use that to fix the issue :)

**Reply by cdietrich** [reply]
@msujew i can see the problem with 2.1.2 too. i have updated the topic with the grammar and sample model. i invoke completion after the `M`

**Reply by msujew** [reply]
I see, that makes sense. I'll take a look at it and probably release a new patch version afterwards. I think I know what's wrong.

**Reply by msujew** [reply]
FYI, I've identified the issue - it's just fairly complicated to solve. I'll be gone during the next week, but hopefully should be able to work on a fix once I'm back.

In the meantime: I guess you can figure out whether an element is synthetic by looking at the missing `$cstNode` property. And just use the `$container` type when it makes sense.

**Reply by lars-reimann** [reply]
> Note that the AST picking logic might still be riddled with edge case issues, so a reproducible example would be great so we can use that to fix the issue :)

Would further examples still be helpful? Then I'd try to minimize some examples I ran into (v3.0.0).

**Reply by cdietrich** [reply]
@msujew as we ran into this again? would a issue be helpful?

**Reply by msujew** [reply]
@cdietrich Sure, please create one.

**Reply by cdietrich** [reply]
created https://github.com/eclipse-langium/langium/issues/1581
